unstable/propertyvalue/propertyvalue.go

```diff
@@ -56,13 +58,49 @@ func TransformPropertyValue(
 	transformer func(resource.PropertyPath, resource.PropertyValue) (resource.PropertyValue, error),
 	value resource.PropertyValue,
 ) (resource.PropertyValue, error) {
+	return TransformPropertyValueDirectional(
+		path, func(path resource.PropertyPath, value resource.PropertyValue, entering bool) (resource.PropertyValue, error) {
+			if !entering {
+				return transformer(path, value)
+			}
+			return value, nil
+		}, value,
+	)
+}
+
+type SkipChildrenError struct{}
+
+func (SkipChildrenError) Error() string {
+	return "skip children"
+}
+
+type TransformerDirectional = func(
+	resource.PropertyPath, resource.PropertyValue, bool,
+) (resource.PropertyValue, error)
+
+// TransformPropertyValueDirectional is a variant of TransformPropertyValue that allows the transformer
+// to visit nodes both on the way down and on the way up.
+//
+// The transformer can return a SkipChildrenError to indicate that the recursion should not descend into the value.
+func TransformPropertyValueDirectional(
```

Comment:
I'd rather this be called `TransformPropertyValue`, but we can change that in a subsequent PR.